src/utils.zig

```diff
@@ -1530,6 +1530,195 @@ pub const BinaryWriter = struct {
 	}
 };
 
+const ReadWriteTest = struct {
+	var testingAllocator = main.heap.ErrorHandlingAllocator.init(std.testing.allocator);
+	var allocator = testingAllocator.allocator();
+
+	fn getWriter() BinaryWriter {
+		return .init(ReadWriteTest.allocator);
+	}
+	fn getReader(data: []const u8) BinaryReader {
+		return .init(data);
+	}
+	fn testInt(comptime IntT: type, expected: IntT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeInt(IntT, expected);
+
+		const expectedWidth = std.math.divCeil(comptime_int, @bitSizeOf(IntT), 8);
+		try std.testing.expectEqual(expectedWidth, writer.data.items.len);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readInt(IntT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn testFloat(comptime FloatT: type, expected: FloatT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeFloat(FloatT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readFloat(FloatT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn testEnum(comptime EnumT: type, expected: EnumT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeEnum(EnumT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readEnum(EnumT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn TestEnum(comptime IntT: type) type {
+		return enum(IntT) {
+			first = std.math.minInt(IntT),
+			center = (std.math.maxInt(IntT) + std.math.minInt(IntT))/2,
+			last = std.math.maxInt(IntT),
+		};
+	}
+	fn testVec(comptime VecT: type, expected: VecT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeVec(VecT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readVec(VecT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+};
+
+test "read/write unsigned int" {
+	inline for([_]type{u8, u1, u2, u4, u5, u8, u16, u31, u32, u64, u128}) |intT| {
```

Comment:
`u8` is listed twice, also would it be possible to add `u0`? It would be interesting to see how it behaves in the edge case of zero-sized types.

Reply:
That is definitely a typo

Reply:
Well, so behavior is basically that nothing is stored and nothing is red, which is kind of valid, but not very useful, is it?

Reply:
No it isn't, but it is an edge case that makes sense testing in my opinion.

Reply:
`u0` can be e.g. the backing type of enums with only one entry.